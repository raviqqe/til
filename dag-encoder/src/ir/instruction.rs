use crate::Operand;
#[cfg(feature = "alloc")]
use alloc::vec::Vec;

/// An instruction.
#[derive(Debug, Eq, PartialEq)]
pub enum Instruction {
    /// A `constant` instruction.
    Constant(Operand),
    /// A `get` instruction.
    Get(Operand),
    /// A `set` instruction.
    Set(Operand),
    /// An `if` instruction.
    #[cfg(feature = "alloc")]
    If(Vec<Instruction>),
    /// A `nop` instruction.
    Nop(u64),
    /// A `call` instruction.
    Call(u64, Operand),
    /// A `close` instruction.
    ///
    /// It is used only for encoding.
    #[cfg(feature = "alloc")]
    Close(u64, Vec<Instruction>),
    /// A `skip` instruction.
    ///
    /// It is used only for encoding.
    Skip(u64),
}

impl Instruction {
    /// A `constant` instruction.
    pub const CONSTANT: u8 = 0;
    /// A `get` instruction.
    pub const GET: u8 = 1;
    /// A `set` instruction.
    pub const SET: u8 = 2;
    /// An `if` instruction.
    pub const IF: u8 = 3;
    /// A `nop` instruction.
    pub const NOP: u8 = 4;
    /// A `call` instruction.
    pub const CALL: u8 = 5;
    /// A `close` instruction.
    pub const CLOSE: u8 = 6;
    /// A `skip` instruction.
    pub const SKIP: u8 = 7;
}

#[cfg(feature = "alloc")]
mod display {
    use super::*;
    use core::fmt::{self, Display, Formatter};

    impl Instruction {
        /// Displays instructions in a slice.
        pub fn display_slice(instructions: &[Self]) -> impl Display + '_ {
            DisplayInstructionList::new(instructions, 0)
        }
    }

    struct DisplayInstruction<'a> {
        instruction: &'a Instruction,
        #[allow(unused)]
        indent: usize,
    }

    impl<'a> DisplayInstruction<'a> {
        const fn new(instruction: &'a Instruction, indent: usize) -> Self {
            Self {
                instruction,
                indent,
            }
        }
    }

    impl<'a> Display for DisplayInstruction<'a> {
        fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
            let indent = self.indent + 1;

            write!(formatter, "- ")?;

            match self.instruction {
                Instruction::Constant(operand) => write!(formatter, "constant {operand}"),
                Instruction::Get(operand) => write!(formatter, "get {operand}"),
                Instruction::Set(operand) => write!(formatter, "set {operand}"),
                Instruction::If(instructions) => {
                    write!(formatter, "if")?;
                    write!(
                        formatter,
                        "{}",
                        DisplayInstructionList::new(instructions, indent)
                    )
                }
                Instruction::Nop(operand) => write!(formatter, "nop {operand}"),
                Instruction::Call(arity, operand) => {
                    write!(formatter, "call {arity} {operand}")
                }
                Instruction::Close(arity, instructions) => {
                    write!(formatter, "close {arity}")?;
                    write!(
                        formatter,
                        "{}",
                        DisplayInstructionList::new(instructions, indent)
                    )
                }
                Instruction::Skip(count) => write!(formatter, "skip {count}"),
            }
        }
    }

    struct DisplayInstructionList<'a> {
        instructions: &'a [Instruction],
        indent: usize,
    }

    impl<'a> DisplayInstructionList<'a> {
        const fn new(instructions: &'a [Instruction], indent: usize) -> Self {
            Self {
                instructions,
                indent,
            }
        }
    }

    impl<'a> Display for DisplayInstructionList<'a> {
        fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
            for instruction in self.instructions {
                writeln!(formatter)?;

                for _ in 0..self.indent {
                    write!(formatter, "  ")?
                }

                write!(
                    formatter,
                    "{}",
                    DisplayInstruction::new(instruction, self.indent)
                )?;
            }

            Ok(())
        }
    }
}
