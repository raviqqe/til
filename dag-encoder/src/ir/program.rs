use super::instruction::Instruction;
use alloc::{borrow::ToOwned, string::String, vec::Vec};
use core::fmt::{self, Display, Formatter};

const ESCAPED_SIGNS: &[&str] = &["\\", "+", "*", "_"];

/// A program on a virtual machine.
#[derive(Debug, Eq, PartialEq)]
pub struct Program {
    symbols: Vec<String>,
    instructions: Vec<Instruction>,
}

impl Program {
    /// Creates a program.
    pub const fn new(symbols: Vec<String>, instructions: Vec<Instruction>) -> Self {
        Self {
            symbols,
            instructions,
        }
    }

    /// Returns symbols in a program.
    pub fn symbols(&self) -> &[String] {
        &self.symbols
    }

    /// Returns instructions in a program.
    pub fn instructions(&self) -> &[Instruction] {
        &self.instructions
    }
}

impl Display for Program {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        writeln!(formatter, "# symbols")?;

        for symbol in &self.symbols {
            let mut symbol = symbol.clone();

            for sign in ESCAPED_SIGNS {
                symbol = symbol.replace(sign, &("\\".to_owned() + sign));
            }

            writeln!(formatter, "- {symbol}")?;
        }

        write!(formatter, "# instructions")?;
        write!(
            formatter,
            "{}",
            Instruction::display_slice(&self.instructions)
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Operand;
    use alloc::{format, vec};
    use insta::assert_snapshot;

    #[test]
    fn display_symbols() {
        assert_snapshot!(Program::new(vec!["foo".into(), "bar".into()], vec![],));
    }

    #[test]
    fn display_symbols_with_special_signs() {
        for &sign in ESCAPED_SIGNS {
            assert_snapshot!(
                sign.replace('*', "star"),
                Program::new(vec![format!("{}", sign)], vec![])
            );
        }
    }

    #[test]
    fn display_if() {
        assert_snapshot!(Program::new(
            vec![],
            vec![Instruction::If(vec![Instruction::Constant(
                Operand::Integer(42)
            )])],
        ));
    }

    #[test]
    fn display_closure() {
        assert_snapshot!(Program::new(
            vec![],
            vec![Instruction::Close(
                42,
                vec![Instruction::Constant(Operand::Integer(2045))],
            )],
        ));
    }

    #[test]
    fn display_closure_with_if() {
        assert_snapshot!(Program::new(
            vec![],
            vec![
                Instruction::Constant(Operand::Integer(0)),
                Instruction::Constant(Operand::Integer(1)),
                Instruction::Close(
                    42,
                    vec![
                        Instruction::Constant(Operand::Integer(2)),
                        Instruction::If(vec![
                            Instruction::Constant(Operand::Integer(3)),
                            Instruction::Constant(Operand::Integer(4)),
                        ]),
                        Instruction::Constant(Operand::Integer(5)),
                        Instruction::Constant(Operand::Integer(6)),
                    ],
                ),
                Instruction::Constant(Operand::Integer(7)),
                Instruction::Constant(Operand::Integer(8)),
            ],
        ));
    }
}
