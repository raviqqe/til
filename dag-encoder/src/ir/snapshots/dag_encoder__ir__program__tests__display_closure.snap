---
source: dag-encoder/src/ir/program.rs
expression: "Program::new(vec![],\nvec![Instruction::Close(42,\nvec![Instruction::Constant(Operand::Integer(2045))],)],)"
---
# symbols
# instructions
- close 42
  - constant i2045
